### ISSUE TITLE
warn when using `impl Drop` values in methods like `.then_some`

### ISSUE THREAD

**Issue by JasmineLowen** [rust-lang/rust-clippy#16506]
### What it does

I had some really nasty bug in a private code base recently:

```rs
let successfully_aquired_custom_lock = some_predicate();
successfully_aquired_custom_lock.them_some(CustomLock(..));
```

where I had a custom `Drop` impl for `CustomLock`. This will always trigger the custom `Drop` impl, which I forgot. In this particular case, it was always freeing up a lock even in the branch of the code where the predicate said it wasn't available. This lead to some nasty concurrency bugs.

After some time I remembered the evaluation order regarding `them_some` and solved the bug, but this was a lot of time spent for something that could've been caught (easily?).

I don't know how easy it is for clippy to know about types and respective values implementing `Drop` but I would suggest that a custom `Drop` impl should trigger a warning for methods like:

- `.them_some` in `bool`
- `.or` in `Option` and `Result`
- `.ok_or` in `Option` and (soon) `bool`
- `.unwrap_or` in `Option` and `Result`
- `.map_or` in `Option` and `Result`
- `or_insert` on the `Entry` APIs of Maps/Trees etc
- probably more?

So that users can opt in and if they do, they have to justify not using the closure for all of these cases, leaving reasons in the code why it's fine to drop immediately.

I think this would be a good improvement for code bases that are heavily relying on locks and custom locks in specific.

### Advantage

- prevent bugs that unlock locks in code branches where they shouldn't
- more awareness and education about the value- vs. closure-level APIs

### Drawbacks

- implementation work
- otherwise none? I'd suggest to make this opt-in so there should be no performance or verbosity drawbacks for regular users

### Example

```rs
let successfully_aquired_custom_lock = some_predicate();
successfully_aquired_custom_lock.them_some(CustomLock(..));
```

Could be written as:

```rs
let successfully_aquired_custom_lock = some_predicate();
successfully_aquired_custom_lock.them(|| CustomLock(..));
```

goes well with a message like:

`using value with custom `Drop` impl inside of non-lazy function '.them_some'. Note that this will always trigger the `Drop` impl which might be undesirable!`

**Comment by plasmaDestroyer** [comment]
I'm interested in implementing this lint. Has there been any discussion about whether this would be accepted? Are there any existing lints I should look at as reference implementations?
